File: examples/vault-service-auth.js
/**
 * Vault Service Authentication Example using Keycloak JWT (Node.js/JavaScript)
 * 
 * This example demonstrates how Node.js services can authenticate with HashiCorp Vault
 * using Keycloak-issued JWT tokens for tenant-scoped access to secrets.
 */

const axios = require('axios');

class VaultJWTClient {
    constructor(vaultConfig = {}, keycloakConfig = {}) {
        this.vaultConfig = {
            url: 'http://localhost:8200',
            jwtPath: 'auth/jwt/login',
            jwtRole: 'matric-service',
            ...vaultConfig
        };
        
        this.keycloakConfig = {
            url: 'http://localhost:8081',
            realm: 'matric-dev',
            clientId: 'matric-platform',
            clientSecret: process.env.KEYCLOAK_CLIENT_SECRET || 'dev-secret',
            ...keycloakConfig
        };
        
        this.vaultToken = null;
        this.tokenExpiresAt = null;
        
        // Create axios instance with default config
        this.httpClient = axios.create({
            timeout: 10000,
            headers: {
                'Content-Type': 'application/json'
            }
        });
    }

    /**
     * Obtain a JWT token from Keycloak with custom claims.
     * 
     * In a real implementation, this would be done through:
     * 1. Service account authentication
     * 2. Token exchange with custom claims
     * 3. Or direct client credentials flow with custom mappers
     */
    async getKeycloakToken(tenantId, userId) {
        const tokenUrl = `${this.keycloakConfig.url}/realms/${this.keycloakConfig.realm}/protocol/openid_connect/token`;
        
        const data = new URLSearchParams({
            grant_type: 'client_credentials',
            client_id: this.keycloakConfig.clientId,
            client_secret: this.keycloakConfig.clientSecret,
            scope: 'openid profile'
        });
        
        try {
            const response = await this.httpClient.post(tokenUrl, data, {
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded'
                }
            });
            
            return response.data.access_token;
        } catch (error) {
            throw new Error(`Failed to get Keycloak token: ${error.message}`);
        }
    }

    /**
     * Authenticate with Vault using JWT token and get Vault token.
     */
    async authenticateWithVault(jwtToken) {
        const authUrl = `${this.vaultConfig.url}/v1/${this.vaultConfig.jwtPath}`;
        
        const payload = {
            role: this.vaultConfig.jwtRole,
            jwt: jwtToken
        };
        
        try {
            const response = await this.httpClient.post(authUrl, payload);
            const authData = response.data;
            
            // Store vault token and expiration
            this.vaultToken = authData.auth.client_token;
            const leaseDuration = authData.auth.lease_duration;
            this.tokenExpiresAt = Date.now() + (leaseDuration * 1000) - 60000; // Refresh 1 minute early
            
            return authData;
        } catch (error) {
            throw new Error(`Failed to authenticate with Vault: ${error.message}`);
        }
    }

    /**
     * Ensure we have a valid Vault token, refreshing if necessary.
     */
    async ensureAuthenticated(tenantId, userId) {
        if (!this.vaultToken || !this.tokenExpiresAt || Date.now() >= this.tokenExpiresAt) {
            console.log(`Obtaining new Vault token for tenant ${tenantId}, user ${userId}`);
            const jwtToken = await this.getKeycloakToken(tenantId, userId);
            await this.authenticateWithVault(jwtToken);
        }
    }

    /**
     * Retrieve a secret from Vault with automatic authentication.
     */
    async getSecret(path, tenantId, userId) {
        await this.ensureAuthenticated(tenantId, userId);
        
        const secretUrl = `${this.vaultConfig.url}/v1/secret/data/${path}`;
        
        try {
            const response = await this.httpClient.get(secretUrl, {
                headers: {
                    'X-Vault-Token': this.vaultToken
                }
            });
            
            return response.data;
        } catch (error) {
            if (error.response?.status === 403) {
                throw new Error(`Access denied to secret: ${path}`);
            }
            throw new Error(`Failed to get secret: ${error.message}`);
        }
    }

    /**
     * Store a secret in Vault with automatic authentication.
     */
    async putSecret(path, data, tenantId, userId) {
        await this.ensureAuthenticated(tenantId, userId);
        
        const secretUrl = `${this.vaultConfig.url}/v1/secret/data/${path}`;
        const payload = { data };
        
        try {
            const response = await this.httpClient.post(secretUrl, payload, {
                headers: {
                    'X-Vault-Token': this.vaultToken
                }
            });
            
            return response.data;
        } catch (error) {
            if (error.response?.status === 403) {
                throw new Error(`Access denied to store secret: ${path}`);
            }
            throw new Error(`Failed to store secret: ${error.message}`);
        }
    }

    /**
     * List secrets at a given path with automatic authentication.
     */
    async listSecrets(path, tenantId, userId) {
        await this.ensureAuthenticated(tenantId, userId);
        
        const listUrl = `${this.vaultConfig.url}/v1/secret/metadata/${path}`;
        
        try {
            const response = await this.httpClient.get(listUrl, {
                headers: {
                    'X-Vault-Token': this.vaultToken
                },
                params: {
                    list: true
                }
            });
            
            return response.data;
        } catch (error) {
            if (error.response?.status === 403) {
                throw new Error(`Access denied to list secrets: ${path}`);
            }
            throw new Error(`Failed to list secrets: ${error.message}`);
        }
    }
}

class TenantService {
    constructor(tenantId, userId) {
        this.tenantId = tenantId;
        this.userId = userId;
        this.vaultClient = new VaultJWTClient();
    }

    /**
     * Get tenant-specific configuration.
     */
    async getTenantConfig() {
        try {
            const result = await this.vaultClient.getSecret(
                `tenants/${this.tenantId}/config`,
                this.tenantId,
                this.userId
            );
            return result.data.data;
        } catch (error) {
            console.error(`Error getting tenant config: ${error.message}`);
            return {};
        }
    }

    /**
     * Get tenant-specific database configuration.
     */
    async getTenantDatabaseConfig() {
        try {
            const result = await this.vaultClient.getSecret(
                `tenants/${this.tenantId}/database`,
                this.tenantId,
                this.userId
            );
            return result.data.data;
        } catch (error) {
            console.error(`Error getting tenant database config: ${error.message}`);
            return {};
        }
    }

    /**
     * Get user-specific preferences.
     */
    async getUserPreferences() {
        try {
            const result = await this.vaultClient.getSecret(
                `users/${this.userId}/preferences`,
                this.tenantId,
                this.userId
            );
            return result.data.data;
        } catch (error) {
            console.error(`Error getting user preferences: ${error.message}`);
            return {};
        }
    }

    /**
     * Store a user-specific API key.
     */
    async storeUserApiKey(serviceName, apiKey) {
        try {
            await this.vaultClient.putSecret(
                `users/${this.userId}/api-keys`,
                { [serviceName]: apiKey },
                this.tenantId,
                this.userId
            );
            return true;
        } catch (error) {
            console.error(`Error storing user API key: ${error.message}`);
            return false;
        }
    }

    /**
     * Get common configuration (available to all services).
     */
    async getCommonConfig() {
        try {
            const result = await this.vaultClient.getSecret(
                'common/config',
                this.tenantId,
                this.userId
            );
            return result.data.data;
        } catch (error) {
            console.error(`Error getting common config: ${error.message}`);
            return {};
        }
    }

    /**
     * Initialize database connection using tenant-specific configuration.
     */
    async initializeDatabaseConnection() {
        const dbConfig = await this.getTenantDatabaseConfig();
        
        if (!dbConfig.host) {
            throw new Error('Database configuration not found');
        }
        
        // Example database connection setup
        console.log(`Connecting to database: ${dbConfig.host}`);
        console.log(`Username: ${dbConfig.username}`);
        // Don't log passwords in production
        console.log('Password: ***');
        
        return {
            host: dbConfig.host,
            username: dbConfig.username,
            password: dbConfig.password
        };
    }

    /**
     * Get service configuration with fallback to common config.
     */
    async getServiceConfig() {
        const [tenantConfig, commonConfig] = await Promise.all([
            this.getTenantConfig(),
            this.getCommonConfig()
        ]);
        
        return {
            ...commonConfig,
            ...tenantConfig
        };
    }
}

// Example usage
async function main() {
    const tenantId = 'tenant-001';
    const userId = 'user-123';
    
    console.log(`Initializing service for tenant ${tenantId}, user ${userId}`);
    
    const service = new TenantService(tenantId, userId);
    
    try {
        // Get tenant configuration
        console.log('\n--- Tenant Configuration ---');
        const tenantConfig = await service.getTenantConfig();
        console.log(JSON.stringify(tenantConfig, null, 2));
        
        // Get tenant database configuration
        console.log('\n--- Tenant Database Configuration ---');
        const dbConfig = await service.getTenantDatabaseConfig();
        // Don't print passwords in real applications
        const safeDbConfig = Object.fromEntries(
            Object.entries(dbConfig).map(([k, v]) => [k, k === 'password' ? '***' : v])
        );
        console.log(JSON.stringify(safeDbConfig, null, 2));
        
        // Get user preferences
        console.log('\n--- User Preferences ---');
        const userPrefs = await service.getUserPreferences();
        console.log(JSON.stringify(userPrefs, null, 2));
        
        // Get common configuration
        console.log('\n--- Common Configuration ---');
        const commonConfig = await service.getCommonConfig();
        console.log(JSON.stringify(commonConfig, null, 2));
        
        // Store a user API key
        console.log('\n--- Storing User API Key ---');
        const success = await service.storeUserApiKey('external_service', 'sk_new_api_key_12345');
        console.log(`API key stored: ${success}`);
        
        // Get combined service configuration
        console.log('\n--- Combined Service Configuration ---');
        const serviceConfig = await service.getServiceConfig();
        console.log(JSON.stringify(serviceConfig, null, 2));
        
        // Initialize database connection
        console.log('\n--- Database Connection ---');
        const dbConnection = await service.initializeDatabaseConnection();
        console.log('Database connection initialized');
        
        console.log('\n--- Testing Cross-Tenant Access (Should Fail) ---');
        // Try to access another tenant's secrets (should fail)
        const otherService = new TenantService('tenant-002', userId);
        const otherTenantConfig = await otherService.getTenantConfig();
        console.log(JSON.stringify(otherTenantConfig, null, 2));
        
    } catch (error) {
        console.error(`Error: ${error.message}`);
    }
}

// Export classes for use as modules
module.exports = {
    VaultJWTClient,
    TenantService
};

// Run example if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

File: examples/vault-service-auth.py
#!/usr/bin/env python3
"""
Vault Service Authentication Example using Keycloak JWT

This example demonstrates how services can authenticate with HashiCorp Vault
using Keycloak-issued JWT tokens for tenant-scoped access to secrets.
"""

import os
import json
import requests
import time
from typing import Dict, Any, Optional
from dataclasses import dataclass


@dataclass
class VaultConfig:
    """Vault configuration settings"""
    url: str = "http://localhost:8200"
    jwt_path: str = "auth/jwt/login"
    jwt_role: str = "matric-service"


@dataclass
class KeycloakConfig:
    """Keycloak configuration settings"""
    url: str = "http://localhost:8081"
    realm: str = "matric-dev"
    client_id: str = "matric-platform"
    client_secret: str = ""  # Set from environment
    username: str = ""  # Service account username
    password: str = ""  # Service account password


class VaultJWTClient:
    """
    Vault client that authenticates using Keycloak JWT tokens
    and provides tenant-scoped access to secrets.
    """
    
    def __init__(self, vault_config: VaultConfig, keycloak_config: KeycloakConfig):
        self.vault_config = vault_config
        self.keycloak_config = keycloak_config
        self.vault_token: Optional[str] = None
        self.token_expires_at: Optional[float] = None
        self.session = requests.Session()
    
    def get_keycloak_token(self, tenant_id: str, user_id: str) -> str:
        """
        Obtain a JWT token from Keycloak with custom claims.
        
        In a real implementation, this would be done through:
        1. Service account authentication
        2. Token exchange with custom claims
        3. Or direct client credentials flow with custom mappers
        """
        token_url = f"{self.keycloak_config.url}/realms/{self.keycloak_config.realm}/protocol/openid_connect/token"
        
        # Example using client credentials flow
        # In production, ensure the client has proper custom claim mappers
        data = {
            "grant_type": "client_credentials",
            "client_id": self.keycloak_config.client_id,
            "client_secret": self.keycloak_config.client_secret,
            "scope": "openid profile"
        }
        
        response = self.session.post(token_url, data=data)
        response.raise_for_status()
        
        token_data = response.json()
        return token_data["access_token"]
    
    def authenticate_with_vault(self, jwt_token: str) -> Dict[str, Any]:
        """
        Authenticate with Vault using JWT token and get Vault token.
        """
        auth_url = f"{self.vault_config.url}/v1/{self.vault_config.jwt_path}"
        
        payload = {
            "role": self.vault_config.jwt_role,
            "jwt": jwt_token
        }
        
        response = self.session.post(auth_url, json=payload)
        response.raise_for_status()
        
        auth_data = response.json()
        
        # Store vault token and expiration
        self.vault_token = auth_data["auth"]["client_token"]
        lease_duration = auth_data["auth"]["lease_duration"]
        self.token_expires_at = time.time() + lease_duration - 60  # Refresh 1 minute early
        
        return auth_data
    
    def ensure_authenticated(self, tenant_id: str, user_id: str) -> None:
        """
        Ensure we have a valid Vault token, refreshing if necessary.
        """
        if (self.vault_token is None or 
            self.token_expires_at is None or 
            time.time() >= self.token_expires_at):
            
            print(f"Obtaining new Vault token for tenant {tenant_id}, user {user_id}")
            jwt_token = self.get_keycloak_token(tenant_id, user_id)
            self.authenticate_with_vault(jwt_token)
    
    def get_secret(self, path: str, tenant_id: str, user_id: str) -> Dict[str, Any]:
        """
        Retrieve a secret from Vault with automatic authentication.
        """
        self.ensure_authenticated(tenant_id, user_id)
        
        secret_url = f"{self.vault_config.url}/v1/secret/data/{path}"
        headers = {"X-Vault-Token": self.vault_token}
        
        response = self.session.get(secret_url, headers=headers)
        response.raise_for_status()
        
        return response.json()
    
    def put_secret(self, path: str, data: Dict[str, Any], tenant_id: str, user_id: str) -> Dict[str, Any]:
        """
        Store a secret in Vault with automatic authentication.
        """
        self.ensure_authenticated(tenant_id, user_id)
        
        secret_url = f"{self.vault_config.url}/v1/secret/data/{path}"
        headers = {"X-Vault-Token": self.vault_token}
        payload = {"data": data}
        
        response = self.session.post(secret_url, json=payload, headers=headers)
        response.raise_for_status()
        
        return response.json()
    
    def list_secrets(self, path: str, tenant_id: str, user_id: str) -> Dict[str, Any]:
        """
        List secrets at a given path with automatic authentication.
        """
        self.ensure_authenticated(tenant_id, user_id)
        
        list_url = f"{self.vault_config.url}/v1/secret/metadata/{path}"
        headers = {"X-Vault-Token": self.vault_token}
        params = {"list": "true"}
        
        response = self.session.get(list_url, headers=headers, params=params)
        response.raise_for_status()
        
        return response.json()


class TenantService:
    """
    Example service that uses Vault for tenant-scoped secret management.
    """
    
    def __init__(self, tenant_id: str, user_id: str):
        self.tenant_id = tenant_id
        self.user_id = user_id
        
        # Initialize Vault client
        vault_config = VaultConfig()
        keycloak_config = KeycloakConfig(
            client_secret=os.getenv("KEYCLOAK_CLIENT_SECRET", "dev-secret"),
        )
        
        self.vault_client = VaultJWTClient(vault_config, keycloak_config)
    
    def get_tenant_config(self) -> Dict[str, Any]:
        """Get tenant-specific configuration."""
        try:
            result = self.vault_client.get_secret(
                f"tenants/{self.tenant_id}/config",
                self.tenant_id,
                self.user_id
            )
            return result["data"]["data"]
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 403:
                print(f"Access denied to tenant {self.tenant_id} config")
                return {}
            raise
    
    def get_tenant_database_config(self) -> Dict[str, Any]:
        """Get tenant-specific database configuration."""
        try:
            result = self.vault_client.get_secret(
                f"tenants/{self.tenant_id}/database",
                self.tenant_id,
                self.user_id
            )
            return result["data"]["data"]
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 403:
                print(f"Access denied to tenant {self.tenant_id} database config")
                return {}
            raise
    
    def get_user_preferences(self) -> Dict[str, Any]:
        """Get user-specific preferences."""
        try:
            result = self.vault_client.get_secret(
                f"users/{self.user_id}/preferences",
                self.tenant_id,
                self.user_id
            )
            return result["data"]["data"]
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 403:
                print(f"Access denied to user {self.user_id} preferences")
                return {}
            raise
    
    def store_user_api_key(self, service_name: str, api_key: str) -> bool:
        """Store a user-specific API key."""
        try:
            self.vault_client.put_secret(
                f"users/{self.user_id}/api-keys",
                {service_name: api_key},
                self.tenant_id,
                self.user_id
            )
            return True
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 403:
                print(f"Access denied to store API key for user {self.user_id}")
                return False
            raise
    
    def get_common_config(self) -> Dict[str, Any]:
        """Get common configuration (available to all services)."""
        try:
            result = self.vault_client.get_secret(
                "common/config",
                self.tenant_id,
                self.user_id
            )
            return result["data"]["data"]
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 403:
                print("Access denied to common config")
                return {}
            raise


def main():
    """
    Example usage of the VaultJWTClient for tenant-scoped access.
    """
    # Example tenant and user IDs
    tenant_id = "tenant-001"
    user_id = "user-123"
    
    print(f"Initializing service for tenant {tenant_id}, user {user_id}")
    
    # Initialize tenant service
    service = TenantService(tenant_id, user_id)
    
    try:
        # Get tenant configuration
        print("\n--- Tenant Configuration ---")
        tenant_config = service.get_tenant_config()
        print(json.dumps(tenant_config, indent=2))
        
        # Get tenant database configuration
        print("\n--- Tenant Database Configuration ---")
        db_config = service.get_tenant_database_config()
        # Don't print passwords in real applications
        safe_db_config = {k: v if k != "password" else "***" for k, v in db_config.items()}
        print(json.dumps(safe_db_config, indent=2))
        
        # Get user preferences
        print("\n--- User Preferences ---")
        user_prefs = service.get_user_preferences()
        print(json.dumps(user_prefs, indent=2))
        
        # Get common configuration
        print("\n--- Common Configuration ---")
        common_config = service.get_common_config()
        print(json.dumps(common_config, indent=2))
        
        # Store a user API key
        print("\n--- Storing User API Key ---")
        success = service.store_user_api_key("external_service", "sk_new_api_key_12345")
        print(f"API key stored: {success}")
        
        print("\n--- Testing Cross-Tenant Access (Should Fail) ---")
        # Try to access another tenant's secrets (should fail)
        other_service = TenantService("tenant-002", user_id)
        other_tenant_config = other_service.get_tenant_config()
        print(json.dumps(other_tenant_config, indent=2))
        
    except requests.exceptions.RequestException as e:
        print(f"Request error: {e}")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()